src/java_bytecode/ci_lazy_methods.cpp

```diff
@@ -1,39 +1,483 @@
+      initial_lazy_methods);
+    method_worklist2.insert(
+      method_worklist2.end(),
+      needed_clinits.begin(),
+      needed_clinits.end());
+  }
+
+  std::set<irep_idt> methods_already_populated;
+  std::vector<const code_function_callt *> virtual_callsites;
 
-#include <string>
+  bool any_new_methods;
+  do
+  {
+    any_new_methods=false;
+    while(!method_worklist2.empty())
+    {
+      std::swap(method_worklist1, method_worklist2);
+      for(const auto &mname : method_worklist1)
+      {
+        if(!methods_already_populated.insert(mname).second)
+          continue;
+        auto findit=lazy_methods.find(mname);
+        if(findit==lazy_methods.end())
+        {
+          debug() << "Skip " << mname << eom;
+          continue;
+        }
+        debug() << "CI lazy methods: elaborate " << mname << eom;
+        const auto &parsed_method=findit->second;
+        // Note this wraps *references* to method_worklist2, needed_classes:
+        ci_lazy_methods_neededt new_lazy_methods(
+          method_worklist2,
+          needed_classes,
+          symbol_table);
+        method_converter(
+          *parsed_method.first, *parsed_method.second, new_lazy_methods);
+        gather_virtual_callsites(
+          symbol_table.lookup(mname).value,
+          virtual_callsites);
+        any_new_methods=true;
+      }
+      method_worklist1.clear();
+    }
 
-/// Notes `method_symbol_name` is referenced from some reachable function, and
-/// should therefore be elaborated.
-/// \par parameters: `method_symbol_name`: method name; must exist in symbol
-///   table.
-void ci_lazy_methodst::add_needed_method(const irep_idt &method_symbol_name)
+    // Given the object types we now know may be created, populate more
+    // possible virtual function call targets:
+
+    debug() << "CI lazy methods: add virtual method targets ("
+            << virtual_callsites.size()
+            << " callsites)"
+            << eom;
+
+    for(const auto &callsite : virtual_callsites)
+    {
+      // This will also create a stub if a virtual callsite has no targets.
+      get_virtual_method_targets(
+        *callsite,
+        needed_classes,
+        method_worklist2,
+        symbol_table);
+    }
+  }
+  while(any_new_methods);
+
+  // Remove symbols for methods that were declared but never used:
+  symbol_tablet keep_symbols;
+
+  for(const auto &sym : symbol_table.symbols)
+  {
+    if(sym.second.is_static_lifetime)
+      continue;
+    if(lazy_methods.count(sym.first) &&
+       !methods_already_populated.count(sym.first))
+    {
+      continue;
+    }
+    if(sym.second.type.id()==ID_code)
+      gather_needed_globals(sym.second.value, symbol_table, keep_symbols);
+    keep_symbols.add(sym.second);
+  }
+
+  debug() << "CI lazy methods: removed "
+          << symbol_table.symbols.size() - keep_symbols.symbols.size()
+          << " unreaclass_hierarchyable methods and globals"
+          << eom;
+
+  symbol_table.swap(keep_symbols);
```

Comment:
I don't know - since not my code would rather leave it unless you think its worth me finding out whether this is safe.